core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestLimitMaxWriterNodesCount.java

```diff
@@ -0,0 +1,184 @@
+                ImmutableMap.of());
+        return queryRunner;
+    }
+
+    @Test
+    public void testPlanWhenInsertToPartitionedTable()
+    {
+        @Language("SQL") String query = "INSERT INTO partitioned_target_table VALUES ('one', 'two'), ('three', 'four')";
+
+        Session session = Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(MAX_WRITER_NODES_COUNT, "2")
+                .setSystemProperty(PREFERRED_WRITE_PARTITIONING_MIN_NUMBER_OF_PARTITIONS, "1")
+                .build();
+
+        // TestLimitMaxWriterNodesCount optimizer rule should fire and set the partitionCount to 2 for remote exchanges
+        assertDistributedPlan(
+                query,
+                session,
+                anyTree(
+                        node(TableWriterNode.class,
+                                project(
+                                exchange(LOCAL,
+                                        exchange(REMOTE, Optional.of(2),
+                                                project(
+                                                        values("column_a", "column_b"))))))));
+    }
+
+    @Test
+    public void testPlanWhenInsertToUnpartitionedTableScaleWritersDisabled()
+    {
+        @Language("SQL") String query = "INSERT INTO unpartitioned_target_table VALUES ('one', 'two')";
+
+        Session session = Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(MAX_WRITER_NODES_COUNT, "2")
+                .setSystemProperty(SCALE_WRITERS, "false")
+                .build();
+
+        // TestLimitMaxWriterNodesCount optimizer rule should fire and set the partitionCount to 2 for remote exchanges
+        assertDistributedPlan(
+                query,
+                session,
+                anyTree(
+                        node(TableWriterNode.class,
+                                exchange(LOCAL, Optional.empty(),
+                                        exchange(REMOTE, Optional.of(2),
+                                                values("column_a", "column_b"))))));
+    }
+
+    @Test
+    public void testPlanWhenInsertToUnpartitionedTableScaleWritersEnabled()
+    {
+        @Language("SQL") String query = "INSERT INTO unpartitioned_target_table VALUES ('one', 'two')";
+
+        Session session = Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(MAX_WRITER_NODES_COUNT, "2")
+                .setSystemProperty(SCALE_WRITERS, "true")
+                .build();
+
+        // TestLimitMaxWriterNodesCount optimizer rule should not fire because scale writers is enabled - no partitioning
+        assertDistributedPlan(
+                query,
+                session,
+                anyTree(
+                        node(TableWriterNode.class,
+                                exchange(LOCAL, Optional.empty(),
+                                        exchange(REMOTE, Optional.empty(),
```

Comment:
Why its empty here?

Reply:
It was in the previous approach. Now, it should be filled.

Reply:
Why empty? We can have it here even though it's not utilized. At least, it'll be visible in the explain plan